homeassistant/components/pi_hole/__init__.py

```diff
@@ -21,90 +21,202 @@
 from .const import (
     DOMAIN,
     CONF_LOCATION,
-    DEFAULT_HOST,
+    CONF_SLUG,
     DEFAULT_LOCATION,
     DEFAULT_NAME,
     DEFAULT_SSL,
     DEFAULT_VERIFY_SSL,
     MIN_TIME_BETWEEN_UPDATES,
     SERVICE_DISABLE,
     SERVICE_DISABLE_ATTR_DURATION,
+    SERVICE_DISABLE_ATTR_NAME,
     SERVICE_ENABLE,
+    SERVICE_ENABLE_ATTR_NAME,
 )
 
+
+def ensure_unique_names_and_slugs(config):
+    """Ensure that each configuration dict contains a unique `name` value."""
+    names = {}
+    slugs = {}
+    for conf in config:
+        if conf[CONF_NAME] not in names and conf[CONF_SLUG] not in slugs:
+            names[conf[CONF_NAME]] = conf[CONF_HOST]
+            slugs[conf[CONF_SLUG]] = conf[CONF_HOST]
+        else:
+            raise vol.Invalid(
+                "Duplicate name '{}' (or slug '{}') for '{}' (already in use by '{}'). Each configured Pi-hole must have a unique name.".format(
+                    conf[CONF_NAME],
+                    conf[CONF_SLUG],
+                    conf[CONF_HOST],
+                    names.get(conf[CONF_NAME], slugs[conf[CONF_SLUG]]),
```

Comment:
I'm using the _default value_ functionality of `get` here, to fallback onto identifying the dupe via slug if it was not triggered on name.